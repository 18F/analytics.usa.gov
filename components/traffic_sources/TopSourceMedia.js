import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../js/lib/renderblock";

function TopSourceMedia({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-session-source-medium-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initTopSourceMediaChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(renderBlock.buildCompactBarChart("session_source_medium"));
      return result;
    };
    initTopSourceMediaChart().catch(console.error);
  }, []);

  return (
    <figure id="chart_session_source_medium" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopSourceMedia.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopSourceMedia;
