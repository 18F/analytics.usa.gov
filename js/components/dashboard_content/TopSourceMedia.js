import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";

/**
 * Retrieves the top source media report from the passed data URL and creates a
 * visualization for the count of users visiting sites with from each type of
 * source media for the current agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopSourceMedia({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-session-source-medium-30-days.json`;
  const ref = useRef(null);
  const [sourceMediaData, setSourceMediaData] = useState(null);

  useEffect(() => {
    const initSourceMediaChart = async () => {
      if (!sourceMediaData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setSourceMediaData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder.buildCompactBarChart(
          ref.current,
          sourceMediaData,
          "session_source_medium",
        );
      }
    };
    initSourceMediaChart().catch(console.error);
  }, [sourceMediaData]);

  return (
    <figure id="chart_session_source_medium" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopSourceMedia.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopSourceMedia;
