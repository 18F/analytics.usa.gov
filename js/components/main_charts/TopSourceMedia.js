import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";

/**
 * Retrieves the top source media report from the passed data URL and creates a
 * visualization for the count of users visiting sites with from each type of
 * source media for the current agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopSourceMedia({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-session-source-medium-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initTopSourceMediaChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(renderBlock.buildCompactBarChart("session_source_medium"));
      return result;
    };
    initTopSourceMediaChart().catch(console.error);
  }, []);

  return (
    <figure id="chart_session_source_medium" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopSourceMedia.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopSourceMedia;
