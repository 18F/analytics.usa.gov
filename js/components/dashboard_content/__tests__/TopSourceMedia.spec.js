import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import TopSourceMedia from "../TopSourceMedia";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("TopSourceMedia", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopSourceMedia dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "top-session-source-medium-30-days",
        query: {},
        meta: {
          name: "Top Session Sources (30 Days)",
          description:
            "Top 20 sources that initiated a session for the last 30 days, measured by visits, for all sites.",
        },
        totals: {
          visits: 1484625337,
          by_session_source_medium: {
            "google / organic": 704457119,
            "(direct) / (none)": 660705226,
            "bing / organic": 48783844,
            "google / cpc": 22696936,
            "yahoo / organic": 10122957,
            "(not set)": 9795115,
            "duckduckgo / organic": 5995170,
            "other / email": 4844244,
            "govdelivery / email": 4158644,
            "expected / email": 2215943,
            "outfordelivery / email": 1920103,
            "bing / cpc": 1261349,
            "google_dv360 / cpc": 1250907,
            "IDme / web": 1240864,
            "baidu / organic": 1191790,
            "delivered / email": 991532,
            "feature / email": 896141,
            "facebook / social": 838767,
            "ecosia.org / organic": 644408,
            "mcp / email": 614278,
          },
        },
        taken_at: "2024-03-11T17:08:18.756Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopSourceMedia dataHrefBase="http://www.example.com/data/" />,
      );
      await waitFor(() => screen.getByText("google / organic"));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopSourceMedia dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
