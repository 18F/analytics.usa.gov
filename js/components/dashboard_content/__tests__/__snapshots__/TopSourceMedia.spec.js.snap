// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopSourceMedia when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <figure
    id="chart_session_source_medium"
  >
    <div
      class="data bar-chart"
    >
      <div
        class="bin"
      >
        <div
          class="label"
        >
          google / organic
        </div>
        <div
          class="value"
        >
          47.7%
        </div>
        <div
          class="bar"
          style="width: 47.7%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          (direct) / (none)
        </div>
        <div
          class="value"
        >
          44.8%
        </div>
        <div
          class="bar"
          style="width: 44.8%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          bing / organic
        </div>
        <div
          class="value"
        >
          3.3%
        </div>
        <div
          class="bar"
          style="width: 3.3%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          google / cpc
        </div>
        <div
          class="value"
        >
          1.5%
        </div>
        <div
          class="bar"
          style="width: 1.5%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          yahoo / organic
        </div>
        <div
          class="value"
        >
          0.7%
        </div>
        <div
          class="bar"
          style="width: 0.7%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          (not set)
        </div>
        <div
          class="value"
        >
          0.7%
        </div>
        <div
          class="bar"
          style="width: 0.7%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          duckduckgo / organic
        </div>
        <div
          class="value"
        >
          0.4%
        </div>
        <div
          class="bar"
          style="width: 0.4%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          other / email
        </div>
        <div
          class="value"
        >
          0.3%
        </div>
        <div
          class="bar"
          style="width: 0.3%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          govdelivery / email
        </div>
        <div
          class="value"
        >
          0.3%
        </div>
        <div
          class="bar"
          style="width: 0.3%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          expected / email
        </div>
        <div
          class="value"
        >
          0.2%
        </div>
        <div
          class="bar"
          style="width: 0.2%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          outfordelivery / email
        </div>
        <div
          class="value"
        >
          0.1%
        </div>
        <div
          class="bar"
          style="width: 0.1%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`TopSourceMedia when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    id="chart_session_source_medium"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;

exports[`TopSourceMedia when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    id="chart_session_source_medium"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;
